File: Code/Python/peak_value_in_array.py
    '''
    Peak Element in an Array in Python
    
    An array element is a peak if it is NOT smaller than its neighbours. For corner elements, we need to consider only one neighbour. 
    There might be multiple peak elements in an array, and the solution should report any peak element.
    
    Implementation: 
    
    Divide and Conquer can be used to find a peak in O(Logn) time. The idea is based on the technique of Binary Search to check if the middle element is the peak element or not. 
    If the middle element is not the peak element, then check if the element on the right side is greater than the middle element then there is always a peak element on the 
    right side. If the element on the left side is greater than the middle element then there is always a peak element on the left side. Form a recursion and the peak element 
    can be found in log n time. 
    

    Algorithm: 

    -Create two variables, l and r, initilize l = 0 and r = n-1
    -Iterate the steps below till l <= r, lowerbound is less than the upperbound
    -Check if the mid value or index mid = (l+r)/2, is the peak element or not, if yes then print the element and terminate.
    -Else if the element on the left side of the middle element is greater then check for peak element on the left side, i.e. update r = mid – 1
    -Else if the element on the right side of the middle element is greater then check for peak element on the right side, i.e. update l = mid + 1
    
    
    '''
  
    # A python3 program to find a peak 
    # element element using divide and conquer
 
    # A binary search based function 
    # that returns index of a peak element
def findPeakUtil(arr, low, high, n):
     
    # Find index of middle element
    # (low + high)/2 
     mid = low + (high - low)/2
     mid = int(mid)
     
    # Compare middle element with its 
    # neighbours (if neighbours exist)
    if ((mid == 0 or arr[mid - 1] <= arr[mid]) and
        (mid == n - 1 or arr[mid + 1] <= arr[mid])):
        return mid
 
 
    # If middle element is not peak and 
    # its left neighbour is greater 
    # than it, then left half must 
    # have a peak element
    elif (mid > 0 and arr[mid - 1] > arr[mid]):
        return findPeakUtil(arr, low, (mid - 1), n)
 
    # If middle element is not peak and
    # its right neighbour is greater
    # than it, then right half must 
    # have a peak element
    else:
        return findPeakUtil(arr, (mid + 1), high, n)
 
 
    # A wrapper over recursive 
    # function findPeakUtil()
def findPeak(arr, n):
 
    return findPeakUtil(arr, 0, n - 1, n)
 
 
    # Driver code
if __name__=='__main__':
  
  arr = list(map(int, input().strip().split()))
  n=len(arr)
  index=findPeak(arr,n)
  print("Peak Element is ", arr[index])
  
  
    '''
    
    Testcases:-
    
    1.
    Input:
    5 10 20 15
    
    Output:
    Peak element is 20
    
    2.
    Input:
    8 9 10 2 5 6
    
    Output:
    Peak element is 10
    
    Complexity:-
    Time Complexity: O(Logn). # In each step our search becomes half. So it can be compared to Binary search, So the time complexity is O(log n)
    Space complexity: O(1).  # No extra space is required, so the space complexity is constant.
    '''
  
  

File: Code/Java/peakvalueinarray.java
/*
An array element is a peak if it is NOT smaller than
its neighbours. For corner elements, we need to
consider only one neighbour.
Here we will Assume Peak element in Array is -11 i.e if peak elemwnt dosent exists then will retuen -1,
then by Liner Search, we will compare (i) with (i+1) & (i-1), if found greater than both the Neighbouring Elemnts
then it will retuen that Element.
*/

package peakval;
import java.util.Scanner;

public class peakvalueinarray {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        System.out.println("Enter Size of Array");
        int n = scan.nextInt();
        if (n>0) {
            System.out.println("Enter " + n + " Elements of array");
            int[] arr = new int[n];
            for (int i = 0; i < n; i++) {
                arr[i] = scan.nextInt();
            }
            int ans = peakval(arr);
            System.out.println();
            System.out.println("Peak Value in Array is : " + ans);
        }else {
            System.out.println("Enter Valid Length of Array");
        }
    }
    public static int peakval(int[] arr){
        if (arr.length >=2){
            if(arr[0]>arr[1]){
                return arr[0];
            }
        }
        int peak = -1;
        for (int i = 1;i<arr.length-1;i++){
            if (arr[i]>arr[i+1] && arr[i]>arr[i-1]){
                peak = arr[i];
                return peak;
            }
        }
        if (arr.length >=2){
            if(arr[arr.length-1]>arr[arr.length-2]){
                return arr[arr.length-1];
            }
        }
        return peak;
    }
}
/*
    Test Cases:
        Input : 5
	        8 9 5 2 4
        Output: 9

        Input : 3
		0 1 0
        Output: 1

        Time Complexity: O(n)
        Space Complexity: O(1)
 */
